src/supautils.c

```diff
@@ -746,28 +762,127 @@ supautils_hook(PROCESS_UTILITY_PARAMS)
 			return;
 		}
 
-		case T_DropStmt:
+		/**
+		 * CREATE POLICY
+		 */
+		case T_CreatePolicyStmt:
 		{
-			DropStmt *stmt = (DropStmt *)utility_stmt;
+			CreatePolicyStmt *stmt = (CreatePolicyStmt *)utility_stmt;
 
 			if (superuser()) {
 				break;
 			}
-			if (privileged_extensions == NULL) {
+			if (!is_current_role_privileged()) {
+				break;
+			}
+			if (privileged_role_allow_policies_on_tables == NULL) {
+				break;
+			}
+
+			if (is_table_range_var_in_list_of_tables_string(stmt->table, privileged_role_allow_policies_on_tables)) {
+				bool already_switched_to_superuser = false;
+
+				switch_to_superuser(privileged_extensions_superuser, &already_switched_to_superuser);
+
+				run_process_utility_hook(prev_hook);
+
+				if (!already_switched_to_superuser) {
+					switch_to_original_role();
+				}
```

Comment:
This type of logic is still a problem for https://github.com/supabase/supautils/issues/79. I have a WIP (https://github.com/supabase/supautils/pull/83) for removing `run_process_utility_hook` usage.

I don't think we should add more of this logic until the refactor is done. Otherwise maintenance cost will keep increasing.

Reply:
It's rather urgent to put this feature in to stop issues like users dropping Storage tables from trickling in, and also to unblock the Realtime team. I don't think the refactor should stop us from adding this.

Reply:
The refactor is not the end goal,  pg 16 compat is. This is gonna make it harder to do.

Reply:
The refactor isn't strictly needed for PG16 - imo we should kaizen this and not use PG16-native features at first, and then slowly move towards using PG16 features. I don't think the codebase is big enough that it warrants a refactor.